### PR TITLE
feat: Add immediate:boolean to RunForkOptions

### PR DESCRIPTION
**Pull request by TylorS**
## Type

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

- Adds `immediate: boolean` to `RunForkOptions` to enable creating a Fiber which uses `fiberRuntime.start()` or `fiberRuntime.resume()`

- Adds `RunCallbackOptions<E, A>` which extends `RunForkOptions` with an `onExit` callback, such that runCallback has access to the same fork options as runFork.

## Related

- Related Issue #
- Closes #

### CODE REVIEW

**Comment by mattiamanzati** [comment]
Bonus, could we please have this also on unsafeRunCallback? It is useful to have the callback eventually be called synchronously if the effect is sync as well

**Comment by TylorS** [comment]
@mattiamanzati Updated in the latest pushes

**Comment by mattiamanzati** [comment]
@TylorS Thanks 🙏
Can't wait for this PR to be accepted and released!

**Comment by mikearnaldi** [comment]
> @mattiamanzati I should be able to add the `immediate` flag there as well. I'm going to remove the `Scope` option though, as it could lead to memory leaks if the surrounding scope is open indefinitely as the Fiber is never removed from the Scope upon exiting

Why would the fiber not being removed upon exit? that should be the behaviour

**Comment by TylorS** [comment]
@mikearnaldi When you do `Scope.addFinalizer(scope, Fiber.interrupt(fiber))`, there is no corresponding API for removing the finalizer from the Scope after the Fiber exits. So if the scope it is being added to lives indefinitely (i.e. the scope of a React component's event listeners), the Fibers will be unable to be garbage collected due to being kept within the open Scope.

**Comment by mikearnaldi** [comment]
> @mikearnaldi When you do `Scope.addFinalizer(scope, Fiber.interrupt(fiber))`, there is no corresponding API for removing the finalizer from the Scope after the Fiber exits. So if the scope it is being added to lives indefinitely (i.e. the scope of a React component's event listeners), the Fibers will be unable to be garbage collected due to being kept within the open Scope.

you're right, I will investigate if adding an api to remove a finaliser may be useful

**Comment by TylorS** [comment]
In the same vein, I'm not positive that `forkIn` or `forkScoped` are entirely safe. They do utilize forked Scope's which allows the Fiber to be garbage collected after completion, but there's a (albeit smaller) memory leak of child scopes if the part scope is kept open indefinitely.

**Comment by mikearnaldi** [comment]
```ts
/** @internal */
export const forkIn = dual<
  (scope: Scope.Scope) => <R, E, A>(self: Effect.Effect<R, E, A>) => Effect.Effect<R, never, Fiber.RuntimeFiber<E, A>>,
  <R, E, A>(self: Effect.Effect<R, E, A>, scope: Scope.Scope) => Effect.Effect<R, never, Fiber.RuntimeFiber<E, A>>
>(
  2,
  (self, scope) =>
    core.uninterruptibleMask((restore) =>
      core.flatMap(scope.fork(executionStrategy.sequential), (child) =>
        pipe(
          restore(self),
          core.onExit((exit) => child.close(exit)),
          fiberRuntime.forkDaemon,
          core.tap((fiber) =>
            child.addFinalizer(() =>
              core.fiberIdWith((fiberId) =>
                Equal.equals(fiberId, fiber.id()) ?
                  core.unit :
                  core.asUnit(core.interruptFiber(fiber))
              )
            )
          )
        ))
    )
)
```

This is safe, it forks a separate scope and that scope is closed as the fiber ends

**Comment by TylorS** [comment]
But closing that child/forked Scope doesn't remove its reference as a Finalizer in the parent Scope when it is closed

**Comment by TylorS** [comment]
I take that back, I found the code that keeps it safe here -https://github.com/Effect-TS/effect/blob/main/packages%2Feffect%2Fsrc%2Finternal%2FfiberRuntime.ts

File: packages/effect/src/Effect.ts [hunk 1]
```diff
@@ -4749,7 +4749,7 @@ export const runFork: <E, A>(effect: Effect<never, E, A>) => Fiber.RuntimeFiber<
  */
 export const runCallback: <E, A>(
   effect: Effect<never, E, A>,
-  onExit?: (exit: Exit.Exit<E, A>) => void
+  options?: Runtime.RunCallbackOptions<E, A> | undefined
```
Comment (tim-smart) [inline packages/effect/src/Effect.ts:4752]: Will need to add a minor changeset for this.

**Review by tim-smart — CHANGES_REQUESTED** [review]

File: packages/effect/src/Effect.ts (hunk 1 above)
Comment (TylorS) [reply packages/effect/src/Effect.ts:4752]: Updated

**Review by tim-smart — APPROVED** [review]

**Comment by tim-smart** [comment]
> Bonus, could we please have this also on unsafeRunCallback? It is useful to have the callback eventually be called synchronously if the effect is sync as well

The current behaviour is to start immediately, so you will see no change with this PR.

**Comment by mattiamanzati** [comment]
> > Bonus, could we please have this also on unsafeRunCallback? It is useful to have the callback eventually be called synchronously if the effect is sync as well
> 
> The current behaviour is to start immediately, so you will see no change with this PR.

Uhm but run callback relies on run fork, so it should be affected as well right?

**Comment by TylorS** [comment]
This PR enables non-immediate starts through `immediate:false` as the default is `true` to avoid breaking changes

**Comment by mattiamanzati** [comment]
> enables

Ohh, I though of it in the completely opposite direction.

**Comment by mikearnaldi** [comment]
Can we add back Scope in the style used by forkIn? also it is worth updating the main runFork method in Effect.ts to take the RunOptions currently only Runtime exposes it

**Comment by TylorS** [comment]
@mikearnaldi PR Updated with Scope [here](https://github.com/Effect-TS/effect/pull/1896/commits/e00349277787453ad0e4425425b7657b3183ad0e) and the additional options to Effect.runFork [here](https://github.com/Effect-TS/effect/pull/1896/commits/ec58bab4ee5f6efefe6eb680653f70980b772801)

File: packages/effect/src/internal/runtime.ts [hunk 2]
```diff
@@ -55,6 +55,15 @@ export const unsafeFork = <R>(runtime: Runtime.Runtime<R>) =>
     runtime.runtimeFlags
   )
 
+  let effect: Effect.Effect<R, E, A> = self
+
+  if (options?.scope) {
```
Comment (tim-smart) [inline packages/effect/src/internal/runtime.ts:60]: Need to add the finalizer :)

Comment (TylorS) [reply packages/effect/src/internal/runtime.ts:61]: 🤦 , updated https://github.com/Effect-TS/effect/pull/1896/commits/8215ce67a8a1b0c22e4de4fdac53bbd7e98ea058

**Comment by tim-smart** [comment]
@TylorS if you rebase that should fix CI

**Comment by TylorS** [comment]
Rebased

**Pull request merged**